extensions/notebook/src/book/bookTocManager.ts

```diff
@@ -190,54 +191,49 @@ export class BookTocManager implements IBookTocManager {
 	 * @param addSection The section that'll be added to the target section. If it's undefined then the target section (findSection) is removed from the table of contents.
 	*/
 	async updateTOC(version: BookVersion, tocPath: string, findSection: JupyterBookSection, addSection?: JupyterBookSection): Promise<void> {
-		const toc = yaml.safeLoad((await fs.readFile(tocPath, 'utf8')));
-		this._tocFiles.set(tocPath, toc);
-		let newToc = new Array<JupyterBookSection>(toc.length);
-		for (const [index, section] of toc.entries()) {
-			let newSection = this.buildTOC(version, section, findSection, addSection);
-			if (newSection) {
-				newToc[index] = newSection;
-			}
+		const tocFile = await fs.readFile(tocPath, 'utf8');
+		this.tableofContents = yaml.safeLoad(tocFile);
+		if (!this._tocFiles.has(tocPath)) {
+			this._tocFiles.set(tocPath, tocFile);
+		}
+		const isModified = this.modifyToc(version, this.tableofContents, findSection, addSection);
+		if (isModified) {
+			await fs.writeFile(tocPath, yaml.safeDump(this.tableofContents, { lineWidth: Infinity, noRefs: true, skipInvalid: true }));
+		} else {
+			throw (new Error(loc.sectionNotFound(findSection.title, tocPath)));
 		}
-		await fs.writeFile(tocPath, yaml.safeDump(newToc, { lineWidth: Infinity, noRefs: true, skipInvalid: true }));
-		this.tableofContents = newToc;
 	}
 
 	/**
-	 * Creates a new table of contents structure containing the added section. This method is only called when we move a section to another section.
+	 * Modify the table of contents read from file. This method is only called when we move a section to another section.
 	 * Since the sections can be arranged in a tree structure we need to look for the section that will be modified in a recursively.
 	 * @param version Version of the book
-	 * @param section The current section that we are iterating
+	 * @param toc The table of contents that will be modified
 	 * @param findSection The section that will be modified.
 	 * @param addSection The section that'll be added to the target section. If it's undefined then the target section (findSection) is removed from the table of contents.
+	 * Returns true if the table of contents is modified, else returns false.
```

Comment:
I would further clarify what the specific scenarios are that it would return false since those seem to be important